solver/src/main/java/org/chocosolver/solver/constraints/nary/nvalue/PropNValue.java

```diff
@@ -0,0 +1,164 @@
+/*
+@author Arthur Godet <arth.godet@gmail.com>
+@since 13/03/2020
+*/
+package org.chocosolver.solver.constraints.nary.nvalue;
+
+import gnu.trove.list.array.TIntArrayList;
+import java.util.Random;
+import org.chocosolver.memory.IStateInt;
+import org.chocosolver.solver.constraints.Propagator;
+import org.chocosolver.solver.constraints.PropagatorPriority;
+import org.chocosolver.solver.exception.ContradictionException;
+import org.chocosolver.solver.variables.IntVar;
+import org.chocosolver.solver.variables.events.IntEventType;
+import org.chocosolver.solver.variables.events.PropagatorEventType;
+import org.chocosolver.util.ESat;
+import org.chocosolver.util.objects.setDataStructures.ISet;
+import org.chocosolver.util.objects.setDataStructures.SetFactory;
+import org.chocosolver.util.objects.setDataStructures.SetType;
+import org.chocosolver.util.tools.ArrayUtils;
+
+public class PropNValue extends Propagator<IntVar> {
+    private IntVar nValue;
+    private final int n;
+    private int[] concernedValues;
+    private IStateInt[] witness;
+    private ISet mandatoryValues;
+    private ISet possibleValues;
+    private TIntArrayList listForRandomPick;
+    private Random rnd = new Random(0);
+
+    public PropNValue(IntVar[] vars, IntVar nvalue) {
+        super(ArrayUtils.concat(vars, nvalue), PropagatorPriority.LINEAR, true);
+        this.nValue = nvalue;
+        n = vars.length;
+        TIntArrayList list = new TIntArrayList();
```

Comment:
Why not using a `TIntHashSet` instead ?

Reply:
Because there is no method sort() on a TIntHashSet, and I think there might be code optimisation to do if the values are sorted.

Reply:
But why to you the values in `concernedValues ` to be sorted ?
Alternatively, you could sort `concernedValues` in place using `Arrays.sort`.